### PR TITLE
Add support for Federated principals

### PR DESCRIPTION
**Pull request by ringods**
Signed-off-by: Ringo De Smet <r.de.smet@napoleongames.be>

### CODE REVIEW

**Comment by ringods** [comment]
@tdpauw why do you insist here to convert single values to an array?

https://github.com/thinkinglabs/aws-iam-policy/blob/main/src/principals/serialiser.ts#L16-L18

The AWS docs contain single string values a number of times:

https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html

So I would like to keep single values where these are possible.

**Comment by tdpauw** [comment]
> @tdpauw why do you insist here to convert single values to an array?
> 
> https://github.com/thinkinglabs/aws-iam-policy/blob/main/src/principals/serialiser.ts#L16-L18
> 
> The AWS docs contain single string values a number of times:
> 
> https://docs.aws.amazon.com/IAM/latest/UserGuide/reference_policies_elements_principal.html
> 
> So I would like to keep single values where these are possible.

@ringods This is a leftover when the library only supported the canonical form of IAM Policy documents i.e. everywhere the spec accepts a single value or an array, the library expected an array for ease of processing. In the meantime, support for single values was added by #7

So, you are right, this conversion from a single value to array can be removed.

File: tests/principals/federated.spec.ts [hunk 1]
```diff
@@ -0,0 +1,16 @@
+import {expect} from 'chai';
+import {FederatedPrincipal} from '../../src/principals/federated';
+
+describe('#ServicePrincipal', function() {
+  describe('#toJSON', function() {
+    const identityProvider = 'www.amazon.com';
+    const policy = new FederatedPrincipal(identityProvider);
+
+    it('should return the AWS principal JSON fragment', function() {
```
Comment (tdpauw) [inline tests/principals/federated.spec.ts:9]: I would suggest modifying the test description. I know it is the same as in `principals/service.spec.ts`. But, I realise the naming wasn't great there too.

```suggestion
it('should return the AWS federated principal JSON fragment', function() {
```

File: src/principals/serialiser.ts [hunk 2]
```diff
@@ -25,3 +37,17 @@ class PrincipalJSONSerialiser {
...
+//     if (typeof(currentResult) === 'string') {
+//       currentResult = [currentResult, value];
+//     } else if (Array.isArray(currentResult)) {
+//       currentResult.push(value);
+//     }
+//   } else {
+//     currentResult = value;
+//   };
+// }
+// result[key] = currentResult;
```
Comment (tdpauw) [inline src/principals/serialiser.ts:53]: I guess this can go away.

```suggestion
```

**Review by tdpauw — APPROVED** [review]
Looking good. Thank you for your contribution.
Just two nitpicking remarks 😁

**Pull request merged**